homeassistant/components/media_source/local_source.py

```diff
@@ -0,0 +1,162 @@
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.util import sanitize_path
+
+from .const import DOMAIN, MEDIA_MIME_TYPES
+from .models import BrowseMedia, MediaSource, MediaSourceItem, PlayMedia
+
+
+@callback
+def async_setup(hass: HomeAssistant):
+    """Set up local media source."""
+    source = LocalSource(hass)
+    hass.data[DOMAIN][DOMAIN] = source
+    hass.http.register_view(LocalMediaView(hass))
+
+
+@callback
+def async_parse_identifier(item: MediaSourceItem) -> Tuple[str, str]:
+    """Parse identifier."""
+    if not item.identifier:
+        source_dir_id = "media"
+        location = ""
+
+    else:
+        source_dir_id, location = item.identifier.lstrip("/").split("/", 1)
+
+    if source_dir_id != "media":
+        raise Unresolvable("Unknown source directory.")
+
+    if location != sanitize_path(location):
+        raise Unresolvable("Invalid path.")
+
+    return source_dir_id, location
+
+
+class LocalSource(MediaSource):
+    """Provide local directories as media sources."""
+
+    name: str = "Local Media"
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize local source."""
+        super().__init__(DOMAIN)
+        self.hass = hass
+
+    @callback
+    def async_full_path(self, source_dir_id, location) -> Path:
+        """Return full path."""
+        return self.hass.config.path("media", location)
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> str:
+        """Resolve media to a url."""
+        source_dir_id, location = async_parse_identifier(item)
+        mime_type, _ = await self.hass.async_add_executor_job(
+            mimetypes.guess_type, self.async_full_path(source_dir_id, location)
+        )
+        return PlayMedia(item.identifier, mime_type)
+
+    async def async_browse_media(
+        self, item: MediaSourceItem, media_types: Tuple[str] = MEDIA_MIME_TYPES
+    ) -> BrowseMedia:
+        """Return media."""
+        try:
+            source_dir_id, location = async_parse_identifier(item)
+        except Unresolvable as err:
+            raise BrowseError(str(err)) from err
+
+        return await self.hass.async_add_executor_job(
+            self._browse_media, source_dir_id, location
+        )
+
+    def _browse_media(self, source_dir_id, location):
+        """Browse media."""
+        full_path = Path(self.hass.config.path("media", location))
+
+        if not full_path.exists():
+            raise BrowseError("Path does not exist.")
+
+        if not full_path.is_dir():
+            raise BrowseError("Path is not a directory.")
+
+        return self._build_item_response(source_dir_id, full_path)
+
+    def _build_item_response(self, source_dir_id: str, path: Path, is_child=False):
+        mime_type, _ = mimetypes.guess_type(str(path))
```

Comment:
The first time we call this, it's a directory. Could we not call it if `is_child == False`

Reply:
Yeah, we could. It's not a slow operation though as it's basically just a dictionary lookup and doesn't do any I/O, so I thought readable code was more important.